packages/block-library/src/gallery/edit.js

```diff
@@ -71,146 +65,152 @@ const MOBILE_CONTROL_PROPS_RANGE_CONTROL = Platform.select( {
-			return newImage.caption;
+	/**
+	 * Determines the image attributes that should be applied to an image block
+	 * after the gallery updates.
+	 *
+	 * The gallery will receive the full collection of images when a new image
+	 * is added. As a result we need to reapply the image's original settings if
+	 * it already existed in the gallery. If the image is in fact new, we need
+	 * to apply the gallery's current settings to the image.
+	 *
+	 * @param  {Object} existingBlock Existing Image block that still exists after gallery update.
+	 * @param  {Object} image         Media object for the actual image.
+	 * @return {Object}               Attributes to set on the new image block.
+	 */
+	function buildImageAttributes( existingBlock, image ) {
+		if ( existingBlock ) {
+			return existingBlock.attributes;
 		}
 
-		return currentImageCaption;
+		return {
+			...pickRelevantMediaFiles( image, sizeSlug ),
+			...getHrefAndDestination( image, linkTo ),
+			...getUpdatedLinkTargetSettings( linkTarget, attributes ),
+			sizeSlug,
+		};
 	}
 
-	function onSelectImages( newImages ) {
-		setAttachmentCaptions(
-			newImages.map( ( newImage ) => ( {
-				// Store the attachmentCaption id as a string for consistency
-				// with the type of the id in the images attribute.
-				id: toString( newImage.id ),
-				caption: newImage.caption,
-			} ) )
+	function onSelectImages( selectedImages ) {
+		const imageArray =
+			Object.prototype.toString.call( selectedImages ) ===
+			'[object FileList]'
+				? Array.from( selectedImages ).map( ( file ) => {
+						if ( ! file.url ) {
+							return pickRelevantMediaFiles( {
+								url: createBlobURL( file ),
+							} );
+						}
+
+						return file;
+				  } )
+				: selectedImages;
+
+		const newImages = imageArray
+			.filter(
+				( file ) => file.url || file.type?.indexOf( 'image/' ) === 0
+			)
+			.map( ( file ) => {
+				if ( ! file.url ) {
+					return pickRelevantMediaFiles( {
+						url: createBlobURL( file ),
+					} );
+				}
+
+				return file;
+			} );
+
+		const newAndExistingImages = uniqBy(
+			concat( newImages, images ),
+			'url'
 		);
-		setAttributes( {
-			images: newImages.map( ( newImage ) => ( {
-				...pickRelevantMediaFiles( newImage, sizeSlug ),
-				caption: selectCaption( newImage, images, attachmentCaptions ),
-				// The id value is stored in a data attribute, so when the
-				// block is parsed it's converted to a string. Converting
-				// to a string here ensures it's type is consistent.
-				id: toString( newImage.id ),
-			} ) ),
-			columns: attributes.columns
-				? Math.min( newImages.length, attributes.columns )
-				: attributes.columns,
+
+		const newBlocks = newAndExistingImages.map( ( image ) => {
+			const existingBlock = find(
+				images,
+				( img ) => img.id === image.id
+			);
+
+			return createBlock( 'core/image', {
+				...buildImageAttributes( existingBlock, image ),
+				id: image.id,
+			} );
 		} );
+		replaceInnerBlocks( clientId, newBlocks );
```

Comment:
Do we need to create a new block for images that are already in the gallery? Can't we use already existing ones from `innerBlocks`? On the mobile side, we use the `clientId` as a key so every call of `onSelectImages` forces the re-mount of all images in the gallery. Will it work if it would be something like this?

```js
const onlyNewImages = differenceBy( newImages, images, 'url' );

const newBlocks = onlyNewImages.map( ( image ) => {
	return createBlock( 'core/image', {
		...buildImageAttributes( undefined, image ),
		id: image.id,
	} );
} );

replaceInnerBlocks( clientId, concat( innerBlockImages, newBlocks ) );
```

Reply:
yeh, this would be a nice optimisation, will  give this a try.

Reply:
This worked - I just had to add an extra filter to remove any deleted existing innerblocks if user is returning from an update of the gallery images in the media library dialog